lib/connectors/memory.js

```diff
@@ -792,6 +792,105 @@ Memory.prototype.replaceOrCreate = function(model, data, options, callback) {
   });
 };
 
+Memory.prototype.upsertWithWhere = function(model,data,options,callback){
+
+var self = this;
+  try {
+
+    var response;
+
+    var element = Object.keys(where)[0];
+
+    var filter = {
+      "where": where
+    };
+
+
+    var response = null;
+    if (Object.keys(data).length === 0)
+      response = false;
+    else
+      response = true;
+
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
```

Comment:
For `upsertWithWhere`, we should not merge things from `where` into `data`. The `where` argument should be the matching criteria while the `data` argument should be the properties to be inserted or updated.